### ISSUE TITLE
Need PETScWrappers::MPI::SparseMatrix::mmult that is similar to TrilinosWrappers::SparseMatrix::mmult and SparseMatrix::mmult

### ISSUE THREAD

**Issue by Jie-Cheng** [dealii/dealii#5647]
Both TrilinosWrappers::SparseMatrix and SparseMatrix have mmult operation that does C = A * diag(V) * B, but PETScWrappers::MPI::SparseMatrix does not. I think it would nice to have it. Besides I actually need it to compute the Schur complement of mass matrix: B*{diag(M)}^{-1}*B^T in my application.

I think the implementation can be like this:

1. make a copy of A, named A_copy
2. A_copy.MatDiagonalScale(NULL, V)
3. MatMatMult(A_copy, B, MAT_INITIAL_MATRIX, PETSC_DEFAULT, C)

**Comment by bangerth** [comment]
Yes, but as pointed out in the email list discussion, there are also other functions that may be more appropriate for the current problem, such as `MatMatMatMult` and the PtAP function.

**Comment by Jie-Cheng** [comment]
But in order to do `MatMatMatMult` or `PtAP`, we need to construct a diagonal matrix from V, rather than copying from a template. How to do that? Is it possible to use V directly without creating a Mat?

**Comment by bangerth** [comment]
Good point. If you really just want to take the diagonal, you would have to create a new matrix object just for the diagonal. I agree that for the special case of a diagonal middle matrix, your approach is probably easier.

**Comment by Jie-Cheng** [comment]
`SparseMatrix::mmult` has an option to promise not to change the sparsity pattern, but `TrilinosWrappers::SparseMatrix` does not. In my application, I can predict the sparsity pattern before doing `mmult`. Do you think it is a good idea to add this option to `PETScWrappers::MatrixBase::mmult`?

Besides, in my approach, a temporary hard copy of A is created. I guess it is kinda inefficient, right?

**Comment by bangerth** [comment]
I would suggest first getting a version to work before starting to think about optimizations. You may find out that the function is not in fact all that expensive, and then the optimization is not worth implementing. Or you may find out that the function is indeed expensive, and then you know for a fact that it's worth your time implementing the optimization.
